### PR TITLE
Add CDC partitioner support

### PR DESCRIPTION
**Pull request by Ponewor**
Fixes: #194 

The following PR refactors partitioner logic and implements CDC Partitioner.
## Pre-review checklist

- [x] I have split my patch into logically separate commits.
- [x] All commit messages clearly explain what they change and why.
- [x] I added relevant tests for new features and bug fixes.
- [x] All commits compile, pass static checks and pass test.
- [x] PR description sums up the changes and reasons why they should be introduced.
- [x] I added appropriate `Fixes:` annotations to PR description.

### CODE REVIEW

File: scylla/src/transport/session.rs [hunk 1]
```diff
@@ -1231,12 +1239,10 @@ fn calculate_token(
         Err(PartitionKeyError::ValueTooLong(values_len)) => {
             return Err(QueryError::BadQuery(BadQuery::ValuesTooLongForKey(
                 values_len,
-                u16::max_value().into(),
+                u16::MAX.into(),
```
Comment (piodul) [inline scylla/src/transport/session.rs:1261]: This change should be placed into a separate commit, and maybe even into separate PR as it doesn't seem related.

**Review by piodul — CHANGES_REQUESTED** [review]
It's quite sad that this implementation performs a lookup in the metadata every time `calculate_token` is called, which is also every time a prepared statement is executed. Per-table partitioners will not change during the lifetime of a table, so please precompute this information when the statement is being prepared.

**Comment by Ponewor** [comment]
> It's quite sad that this implementation performs a lookup in the metadata every time calculate_token is called, which is also every time a prepared statement is executed. Per-table partitioners will not change during the lifetime of a table, so please precompute this information when the statement is being prepared.

I am not sure how this could be achieved. `PreparedStatement`s are generated in `Connection::prepare`. `Connection` has no idea about cluster's metadata so there is no direct way to access the partitioner name. Perhaps it could passed top-down but it would require to break a few public interfaces.

**Comment by piodul** [comment]
> > It's quite sad that this implementation performs a lookup in the metadata every time calculate_token is called, which is also every time a prepared statement is executed. Per-table partitioners will not change during the lifetime of a table, so please precompute this information when the statement is being prepared.
> 
> I am not sure how this could be achieved. `PreparedStatement`s are generated in `Connection::prepare`. `Connection` has no idea about cluster's metadata so there is no direct way to access the partitioner name. Perhaps it could passed top-down but it would require to break a few public interfaces.

I think it will be sufficient if we do it in `Session::prepare`. There, you should have all the necessary information and can set the partitioner field before returning `PreparedStatement`.

The `Connection::prepare` can keep returning `PreparedStatement` without the partitioner set. Users are supposed to mainly use `Session` to interact with the cluster, and the partitioner is just an optimization. I think that in the future we will hide the `Connection` from the public interface anyway, and if not, let's not worry about setting partitioner there now.

File: scylla/src/statement/prepared_statement.rs [hunk 2]
```diff
@@ -203,6 +213,16 @@ impl PreparedStatement {
...
+
+    /// Sets the name of the partitioner used for this statement.
+    pub fn set_partitioner_name(&mut self, partitioner_name: Option<String>) {
+        self.partitioner_name = partitioner_name
+    }
+
+    /// Get the name of the partitioner used for this statement.
+    pub fn get_partitioner_name(&self) -> Option<String> {
+        self.partitioner_name.clone()
+    }
```
[the remark this answers is not in the document]
Comment (piodul) [reply scylla/src/statement/prepared_statement.rs:225]: Also, in general, please avoid returning `String` when you can just return `&str`... Sometimes the caller does not want the full copy and this lets avoid them do this. The caller can just always call `to_owned()` if they want a `String`.

This comment probably becomes obsolete if you switch to using a `Partitioner` enum or something similar.

**Review by piodul — CHANGES_REQUESTED** [review]
Please also add a test which verifies that a prepared statement has the right partitioner set for CDC/non-CDC table.

File: scylla/src/transport/session.rs [hunk 3]
```diff
@@ -599,6 +599,19 @@ impl Session {
                 .extend(statement.prepare_tracing_ids);
         }
 
+        let keyspace = prepared.get_keyspace_name();
+        let table = prepared.get_table_name();
+        if keyspace.is_some() && table.is_some() {
```
[the remark this answers is not in the document]
Comment (Ponewor) [reply scylla/src/transport/session.rs:604]: This:
```
fn extract_partitioner_name(&self, prepared: &PreparedStatement) -> Option<&str> {
        let cluster_data = self.cluster.get_data();
        cluster_data
            .keyspaces
            .get(prepared.get_keyspace_name()?)?
            .tables
            .get(prepared.get_table_name()?)?
            .partitioner
    }
 ```
does not work - the types do not match (`Option<&str>` vs Option<String>). Is it even possible to cast it? Any usual tricks such as `.as_deref()` end with result referencing data owned by this method.

File: scylla/src/statement/prepared_statement.rs [hunk 4]
```diff
@@ -19,6 +19,7 @@ pub struct PreparedStatement {
     metadata: PreparedMetadata,
     statement: String,
     page_size: Option<i32>,
+    partitioner_name: Option<String>,
```
[the remark this answers is not in the document]
Comment (Ponewor) [reply scylla/src/statement/prepared_statement.rs:22]: Once we have an enum like
```
pub enum PartitionerName {
    Murmur3,
    CDC,
}
```
what should be the type of this field? There are two possibilities:

- `PartitionerName` with implemented `Default` as `PartitionerName::Murmur3`. The drawback is that it hardcodes "the defaultability" of Murmur3 in the "wrong" place - I don't know how to properly express it but it kinda feels wrong.
- `Option<PartitionerName>`. The drawback is that we have to duplicate the logic handling empty partitioner. First we need to map from string to the field itself when statement is prepared with something like
```
self.partitioner_name = Some(match partitioner_name {
            Some(partitioner_name) if partitioner_name.ends_with("CDCPartitioner") => {
                PartitionerName::CDC
            }
            _ => PartitionerName::Murmur3,
        })
```
and when the statement is executed and we need to actually pick the correct partitioner we need to do something like
```
match partitioner_name {
            None | Some(PartitionerName::Murmur3) => Ok(Murmur3Partitioner::hash(partition_key)),
            Some(PartitionerName::CDC) => Ok(CDCPartitioner::hash(partition_key)),
        }
```
The logic is more or less repeated here and this could be avoided if we chose the first aproach.

What do you think?

File: scylla/src/transport/session.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (piodul) [reply scylla/src/transport/session.rs:604]: This works:

```rust
fn extract_partitioner_name<'a>(
    &self,
    prepared: &PreparedStatement,
    cluster_data: &'a ClusterData,
) -> Option<&'a str> {
    cluster_data
        .keyspaces
        .get(prepared.get_keyspace_name()?)?
        .tables
        .get(prepared.get_table_name()?)?
        .partitioner
        .as_ref()
        .map(|s| s.as_ref())
}
```

You need to get inside the optional in order to get a `&str` from `String`, hence the `.map`.

Note that the returned string borrows from the `cluster_data`. You need to provide it from outside and give explicit lifetimes in order for everything to check out, otherwise Rust will rightfully complain that you are returning a reference to an object that might potentially no longer exist.

File: scylla/src/statement/prepared_statement.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (piodul) [reply scylla/src/statement/prepared_statement.rs:22]: I think the first option is fine. The Murmur3 partitioner _is_ the default partitioner for Scylla. If we keep this type as internal for the crate - not `pub`, at most `pub(crate)` - I don't think it hurts making Murmur3 the `Default`.

**Review by piodul — COMMENTED** [review]
> Please also add a test which verifies that a prepared statement has the right partitioner set for CDC/non-CDC table.

The implementation looks good, but the test is still missing.

**Comment by Ponewor** [comment]
> > Please also add a test which verifies that a prepared statement has the right partitioner set for CDC/non-CDC table.
> 
> The implementation looks good, but the test is still missing.

Right, it somehow has slipped my attention.

While trying to complete this task I have found some unexpected difficulties - the `prepared_metadata` in `PreparedStatement` is always empty (`PreparedMetadata { col_count: 0, pk_indexes: [], col_specs: [] }`). This results in `get_keyspace_name` and `get_table_name` returning `None`.

Is this a bug or is this metadata fetched later? If the latter then when does this happen? How to trigger it?

**Comment by piodul** [comment]
> > > Please also add a test which verifies that a prepared statement has the right partitioner set for CDC/non-CDC table.
> > 
> > 
> > The implementation looks good, but the test is still missing.
> 
> Right, it somehow has slipped my attention.
> 
> While trying to complete this task I have found some unexpected difficulties - the `prepared_metadata` in `PreparedStatement` is always empty (`PreparedMetadata { col_count: 0, pk_indexes: [], col_specs: [] }`). This results in `get_keyspace_name` and `get_table_name` returning `None`.
> 
> Is this a bug or is this metadata fetched later? If the latter then when does this happen? How to trigger it?

Could you paste the query string? It should have at least one `?` in order for the metadata to be non-empty.

**Comment by piodul** [comment]
> > > > > Please also add a test which verifies that a prepared statement has the right partitioner set for CDC/non-CDC table.
> > > > 
> > > > 
> > > > The implementation looks good, but the test is still missing.
> > > 
> > > 
> > > Right, it somehow has slipped my attention.
> > > While trying to complete this task I have found some unexpected difficulties - the `prepared_metadata` in `PreparedStatement` is always empty (`PreparedMetadata { col_count: 0, pk_indexes: [], col_specs: [] }`). This results in `get_keyspace_name` and `get_table_name` returning `None`.
> > > Is this a bug or is this metadata fetched later? If the latter then when does this happen? How to trigger it?
> > 
> > 
> > Could you paste the query string? It should have at least one `?` in order for the metadata to be non-empty.
> 
> The table is given by `CREATE TABLE IF NOT EXISTS t (a int primary key) WITH cdc = {{'enabled':true}}` and the queries I've tried are `SELECT a FROM t` `INSERT INTO t (a) VALUES (0)`

Right, your queries need bind markers. `PreparedMetadata` contains information about bind markers, so if you don't have any then the metadata will be empty. Try changing your queries to `SELECT a FROM t WHERE a = ?` and `INSERT INTO t (a) VALUES (?)`.

**Comment by piodul** [comment]
LGTM. Please rebase onto newest main as merging https://github.com/scylladb/scylla-rust-driver/pull/426 seems to have introduced a conflict in the current PR.

**Comment by Ponewor** [comment]
done

**Review by piodul — APPROVED** [review]
